File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode subtreeWithAllDeepest(TreeNode root) {
        return dfs(root).node;
    }
    
    private class Res {
        int depth; TreeNode node;
        Res(int depth, TreeNode node) {
            this.depth = depth;
            this.node = node;
        }
    }
    
    private Res dfs(TreeNode root) {
        if(root == null) return new Res(-1, null);
        Res l = dfs(root.left);
        Res r = dfs(root.right);
        int dl = l.depth; int dr = r.depth;
        TreeNode node = dl == dr ? root : (dl > dr) ? l.node : r.node;
        return new Res(Math.max(dl, dr)+1, node);
    }
}

File: 0865-smallest-subtree-with-all-the-deepest-nodes/0865-smallest-subtree-with-all-the-deepest-nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def subtreeWithAllDeepest(self, root: TreeNode) -> TreeNode:
        def dfs(root):
            if not root: return (-1, None)
            ld, l = dfs(root.left)
            rd, r = dfs(root.right)
            if ld == rd: return(ld+1, root)
            return (ld + 1, l) if ld > rd else (rd + 1, r)
        return dfs(root)[1]
